src/qutip_qip/qir.py

```diff
@@ -0,0 +1,180 @@
+    #: ending in `.bc`).
+    BITCODE = auto()
+    #: Specifies that QIR should be encoded as plain text (typicaly, files
+    #: ending in `.ll`).
+    TEXT = auto()
+    #: Specifies that QIR should be encoded as a PyQIR module object.
+    MODULE = auto()
+
+    @classmethod
+    def ensure(cls, val : Union[Literal["bitcode", "text", "module"], QirFormat]) -> QirFormat:
+        """
+        Given a value, returns a value ensured to be of type `QirFormat`,
+        attempting to convert if needed.
+        """
+        if isinstance(val, cls):
+            return val
+        elif isinstance(val, str):
+            return cls[val.upper()]
+
+        return cls(val)
+
+# Specify return types for each different format, so that IDE tooling and type
+# checkers can resolve the return type based on arguments.
+@overload
+def circuit_to_qir(circuit : QubitCircuit, format : Union[Literal[QirFormat.BITCODE], Literal["bitcode"]], module_name : str) -> bytes: ...
+@overload
+def circuit_to_qir(circuit : QubitCircuit, format : Union[Literal[QirFormat.TEXT], Literal["text"]], module_name : str) -> str: ...
+@overload
+def circuit_to_qir(circuit : QubitCircuit, format : Union[Literal[QirFormat.MODULE], Literal["module"]], module_name : str) -> pqp.QirModule: ...
+
+def circuit_to_qir(circuit, format, module_name = "qutip_circuit"):
+    """
+    Given a circuit acting on qubits, generates a representation of that
+    circuit using Quantum Intermediate Representation (QIR).
+
+    :param circuit: The circuit to be translated to QIR.
+    :param format: The QIR serialization to be used. If `"text"`, returns a
+        plain-text representation using LLVM IR. If `"bitcode"`, returns a
+        dense binary representation ideal for use with other compilation tools.
+        If `"module"`, returns a PyQIR module object that can be manipulated
+        further before generating QIR.
+    :param module_name: The name of the module to be emitted.
+    """
+    # Define as an inner function to make it easier to call from conditional
+    # branches.
+    def append_operation(module: pqg.SimpleModule, builder: pqg.BasicQisBuilder, op: Gate):
+        if op.classical_controls:
+            result = op.classical_controls[0]
+            value = "zero" if op.classical_control_value == 0 else "one"
+            # Pull off the first control and recurse.
+            op_with_less_controls = Gate(**op.__dict__)
+            op_with_less_controls.classical_controls = op_with_less_controls.classical_controls[1:]
+            branch_body = {value: (lambda: append_operation(module, builder, op_with_less_controls))}
+            builder.if_result(module.results[result], **branch_body)
+            return
+
+        if op.controls:
+            if op.name not in ("CNOT", "CZ") or len(op.controls) != 1:
+                raise NotImplementedError(
+                    "Arbitrary controlled quantum operations are not yet supported."
+                )
```

Comment:
It seems that CZ is missing from the list below?

Reply:
Good catch, thanks!